packages/gravity/kube-publishing/src/test-runner.ts

```diff
@@ -0,0 +1,163 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import crypto from 'node:crypto';
+import fs from 'node:fs';
+import os from 'node:os';
+import path from 'node:path';
+
+import { sleep } from '@dxos/async';
+import { log } from '@dxos/log';
+
+import { PublishTestSpec, CacheStatus, EvaluationResult } from './spec';
+import { run } from './utils';
+
+const cfCacheHeader = 'cf-cache-status';
+
+export class TestRunner {
+  private _appPath: string;
+  // Map of file paths to their expected hashes.
+  private _filesMap: Record<string, string> = {};
+  private _testId = Date.now();
+
+  constructor(private readonly _spec: PublishTestSpec) {
+    this._appPath = `${os.tmpdir()}/kube-publishing/${this._spec.appName}`;
```

Comment:
Better to make this path structured even for tmp files.
`/tmp/dxos/kube/publishing`